torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1362,6 +1371,433 @@ def rbinary_magic_impl(self, other):
 del method
 del func
 
+try:
+    import z3
+
+    Z3Expr = Union[z3.ArithRef, z3.BoolRef]
+
+    def _boolref(expr: Any) -> z3.BoolRef:
+        return cast(z3.BoolRef, expr)
+
+    # Wrapper of Z3 expressions.
+    #
+    # Some operations on Z3 depend on an assertion so that Z3 can reason
+    # about it correctly. e.g. division depends on the denominator being
+    # different than 0.
+    #
+    # This class helps propagate such an assertion to the caller.
```

Comment:
For example, `div(a, b)` will only work correctly if we add `b != 0` to Z3 solver. So, we need to propagate these new assertions to the validator. It's basically a named tuple.

Reply:
Could we guard on this whenever we do division so you don't need to track manually?

Reply:
I think it would work for the input guards, when we are creating the FX nodes. But the only way I could see it working for output guards would be to call `self.shape_env.evaluate_expr` from the overridden `SymNode` operations (inside `_make_node_magic`). What do you think?

Reply:
I think it is OK (good, even) to guard on `y != 0` when computing `x / y`.